### ISSUE TITLE
[css-sizing] Does indefinite `stretch` behave as automatic size or as initial value?

### ISSUE THREAD

**Issue by Loirooriol** [w3c/csswg-drafts#11006]
First, consider

```html
<!DOCTYPE html>
<div style="min-height: max-content; height: 0; width: 200px; border: solid">lorem ipsum</div>
```

| Spec, Blink | Gecko, WebKit | 
| - | - |
| ![](https://github.com/user-attachments/assets/257a35dd-2f2e-40bb-8d70-da86787e8153) | ![](https://github.com/user-attachments/assets/45702af0-3ab7-40e2-ad36-90a060e82fe4) |

@bfgeek told me that  Blink had to deal with some breakage when they adopted this behavior, but that it's the correct thing to do. But what about `stretch`?

```html
<!DOCTYPE html>
<div style="min-height: -moz-available; min-height: -webkit-fill-available; min-height: stretch; 
            height: 0; width: 200px; border: solid">lorem ipsum</div>
```

| Spec | Gecko, Blink | WebKit |
| - | - | - |
| ![](https://github.com/user-attachments/assets/257a35dd-2f2e-40bb-8d70-da86787e8153) | ![](https://github.com/user-attachments/assets/45702af0-3ab7-40e2-ad36-90a060e82fe4) | ![](https://github.com/user-attachments/assets/2e946436-f436-4dae-94a4-524812784516) |

(WebKit seems to stretch elements with an indefinite height that contain some `-webkit-fill-available`, so it's always definite)

From https://drafts.csswg.org/css-sizing-4/#stretch-fit-sizing

> Formally, its behavior is the same as specifying an [automatic size](https://drafts.csswg.org/css-sizing-3/#automatic-size) together with a [self-alignment property](https://drafts.csswg.org/css-align-3/#self-alignment-properties) value of [stretch](https://drafts.csswg.org/css-sizing-4/#valdef-width-stretch)

But here Blink is not implementing this, not sure if due to an oversight.

And then the situation is that browsers are implementing `stretch`, but nobody follows the spec here. It may be too late for `*-content`, but I would like to have an agreement for `stretch` before it ships. So Agenda+.

**Comment by Loirooriol** [comment]
For Servo following the spec seems a bit simpler than aligning with Blink, so I guess I will go with that for now.

**Comment by bfgeek** [comment]
> > Formally, its behavior is the same as specifying an [automatic size](https://drafts.csswg.org/css-sizing-3/#automatic-size) together with a [self-alignment property](https://drafts.csswg.org/css-align-3/#self-alignment-properties) value of [stretch](https://drafts.csswg.org/css-sizing-4/#valdef-width-stretch)

So I think this was intended just to be applicable to main sizes, (e.g. not min/max-sizes).

IMO this should be consistent with `min-height: 100%`, e.g. if a percentage is indefinite it behaves as zero for `min-height`, and Infinity for `max-height`.

(e.g. The Blink/Gecko behaviour).

For example it'd be pretty weird if:

```
<!DOCTYPE html>
<div style="max-height: stretch; height: 200px; width: 200px; border: solid">lorem ipsum</div>
```
got clamped by the intrinsic size (when `stretch` is an extrinsic constraint).

cc/ @tabatkins

**Comment by Loirooriol** [comment]
I'm not convinced we should treat it like a percentage. For example, with a min/max-content of 100px, `fit-content` is basically `clamp(100px, stretch, 100px)`, and in Blink `max-height: fit-content` behaves as `100px` but `max-height: clamp(100px, 100%, 100px)` behaves as `none`. And doing this like Blink seems more annoying to implement in Servo.

So I think we should either:
 - Keep the spec as-is, or
 - Treat both `fit-content` and `stretch` as containing a cyclic percentage.
   More specifically, treat `fit-content` as `fit-content(100%)` and `stretch` as `100%`, so:
   - For preferred and max properties they both behave as the initial value ([with some special behavior for replaced elements](https://drafts.csswg.org/css-sizing-3/#cyclic-percentage-contribution))
   - For min properties see #10969

**Comment by bfgeek** [comment]
> I'm not convinced we should treat it like a percentage. For example, with a min/max-content of 100px, `fit-content` is basically `clamp(100px, stretch, 100px)`

This isn't correct - we recently did a change to follow Firefox in this regard. See: https://github.com/w3c/csswg-drafts/issues/10721
When the stretch is indefinite - `min-size: fit-content` resolves to `min-size: min-content`, and `max-size: fit-content` to `max-size: max-content` respectively.

> And doing this like Blink seems more annoying to implement in Servo.

I'm curious why this is difficult in Servo? This might result in difficulty when attempting to implement calc-size() for example.

I'm pretty convinced that we should just be consistent with 100% here, (again see the `max-height:stretch` example above, it'd be super weird if we coerce an explicitly extrinsic constraint to be intrinsic).

**Comment by Loirooriol** [comment]
> we recently did a change to follow Firefox in this regard. See: https://github.com/w3c/csswg-drafts/issues/10721

Not sure if that is the best behavior, I have commented there.

> I'm curious why this is difficult in Servo?

I misunderstood what you were doing for `fit-content`, I thought it was trickier. Resolving keywords in different ways depending on the property and having a potentially indefinite stretch size during layout still adds some complexity. But not a big deal I guess.

> it'd be super weird if we coerce an explicitly extrinsic constraint to be intrinsic

Maybe a bit weird indeed.

**Comment by Loirooriol** [comment]
It's worth noting that there is a difference among major browsers:
 - Gecko treats an indefinite `min-height: stretch` as the initial `auto`
 - Blink and WebKit treat it as `0px`

```html
<!DOCTYPE html>
<div style="display: flex; flex-direction: column; border: solid cyan; min-height: 100px; width: 200px;">
  <div style="min-height: -moz-available; min-height: -webkit-fill-available; min-height: stretch;
              flex-basis: 0; border: solid">lorem ipsum</div>
</div>
```

| Gecko | Blink, WebKit |
| - | - |
| ![](https://github.com/user-attachments/assets/f9de0525-24e1-4f5f-a287-f4c8ed394b8e) | ![](https://github.com/user-attachments/assets/b3c54bfa-0571-45af-bd02-2974ab3014ea) |

Interestingly, they all treat `min-height: 0%` as `0px`.

**Comment by bfgeek** [comment]
I wouldn't read too much into the block direction behaviour with FF (support was lacking last time I checked).

**Comment by dholbert** [comment]
Indeed, the FF behavior is just because I haven't implemented `stretch behavior in the block axis yet (keep getting distracted by other tasks, but made a little local progress this week, hope to be done soonish).

**Comment by Loirooriol** [comment]
Sure, but if the spec changes, we need to decide whether `min-height: stretch` should behave as `auto` or as `0px` or whatever `100%` would do (the spec says that `100%` should behave as `0px`, but [browsers don't follow the spec](https://github.com/w3c/csswg-drafts/issues/10969) in general).

**Comment by davidsgrogan** [comment]
> > > Formally, its behavior is the same as specifying an [automatic size](https://drafts.csswg.org/css-sizing-3/#automatic-size) together with a [self-alignment property](https://drafts.csswg.org/css-align-3/#self-alignment-properties) value of [stretch](https://drafts.csswg.org/css-sizing-4/#valdef-width-stretch)
> 
> So I think this was intended just to be applicable to main sizes, (e.g. not min/max-sizes).
> 
> IMO this should be consistent with `min-height: 100%`, e.g. if a percentage is indefinite it behaves as zero for `min-height`, and Infinity for `max-height`.
> 
> (e.g. The Blink/Gecko behaviour).
> 
> For example it'd be pretty weird if:
> 
> ```
> <!DOCTYPE html>
> <div style="max-height: stretch; height: 200px; width: 200px; border: solid">lorem ipsum</div>
> ```
> 
> got clamped by the intrinsic size (when `stretch` is an extrinsic constraint).
> 
> cc/ @tabatkins

I talked to @tabatkins about this a few weeks ago. He confirmed that the spec authors _intended_ the Blink/Gecko behavior (min-size:stretch resolves to 0 and max-size:stretch resolves to infinity when available size is indefinite) and that it was indeed on the editors to better define the spec around min/max sizes behaving as auto, which should hopefully clear up the confusion here.
